mash_screen
==============

Purpose
-------

This component aims to produce to be a mash screen approach to find plasmids
contained in high throughoput sequencing data. Then, the resulting file can
be imported into .. _pATLAS: http://www.patlas.site/ .
However, this process can be user for other purposes, by providing a different
database than the default that is intended for plasmid searches.

.. note::
    pATLAs software documentation can be found .. _here: https://tiagofilipe12.gitbooks.io/patlas/content/


Input/Output type
------------------

- Input type: ``FastQ``
- Output type: ``JSON``


Parameters
----------

- `noWinner`: A variable that enables the use of -w option for mash screen.
Default: false.

- `pValue`: P-value cutoff for the distance estimation between two sequences to
be included in the output. Default: 0.05.

- `identity`: The percentage of identity between the reads input and the
reference sequence. Default: 0.9.

- `refFile`: "Specifies the reference file to be provided to mash. It can
either be a fasta or a .msh reference sketch generated by mash.
Default: '/ngstools/data/patlas.msh'.


Published results
-----------------

- ``results/mashscreen/``: A `JSON` file that can be imported to .. _pATLAS: http://www.patlas.site/
with the results from mash screen.


Published reports
-----------------

None.


Default directives
------------------

- ``mashScreen``:
    - ``container``: tiagofilipe12/patlasflow_mash_screen
    - ``version``: 1.1
- ``mashOutputJson``:
    - ``container``: tiagofilipe12/patlasflow_mash_screen
    - ``version``: 1.1
