mash_dist
=========

Purpose
-------

This component aims to produce to be a mash dist approach to find plasmids
contained in high throughoput sequencing data. Then, the resulting file can
be imported into .. _pATLAS: http://www.patlas.site/ .
However, this process can be user for other purposes, by providing a different
database than the default that is intended for plasmid searches.

.. note::
    pATLAs software documentation can be found .. _here: https://tiagofilipe12.gitbooks.io/patlas/content/


Input/Output type
------------------

- Input type: ``FastA``
- Output type: ``JSON``


Parameters
----------

- `mash_distance`: Sets the maximum distance between two sequences to be
included in the output. Default: 0.1.

.. note::
    The subtraction of 1 - `mash_distance` can be used as an approximation to
    Average Nucleotide Identity (ANI). For instance a mash distance of 0.1 well
    correlates with ANI at 0.9 (90%).

- `pValue`: P-value cutoff for the distance estimation between two sequences
to be included in the output. Default: 0.05.

- `shared_hashes`: Sets a minimum percentage of hashes shared between two
sequences in order to include its result in the output. Default: 0.8.

- `refFile`: Specifies the reference file to be provided to mash. It can either
be a fasta or a .msh reference sketch generated by mash.
Default: '/ngstools/data/patlas.msh'.


Published results
-----------------

- ``results/mashdist/``: A `JSON` file that can be imported to .. _pATLAS: http://www.patlas.site/
with the results from mash dist.


Published reports
-----------------

None.


Default directives
------------------

- ``runMashDist``:
    - ``container``: tiagofilipe12/patlasflow_mash_screen
    - ``version``: 1.1
- ``mashDistOutputJson``:
    - ``container``: tiagofilipe12/patlasflow_mash_screen
    - ``version``: 1.1
